rust/src/pgsql/pgsql.rs

```diff
@@ -0,0 +1,853 @@
+    let tx_safe: &mut PgsqlTransaction;
+    unsafe {
+        tx_safe = cast_pointer!(tx, PgsqlTransaction);
+    }
+    return tx_safe.tx_state;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_pgsql_tx_get_alstate_progress(
+    tx: *mut std::os::raw::c_void, _direction: u8,
+) -> std::os::raw::c_int {
+    return rs_pgsql_tx_get_state(tx) as i32;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_pgsql_state_get_events(
+    tx: *mut std::os::raw::c_void,
+) -> *mut core::AppLayerDecoderEvents {
+    let tx_safe: &mut PgsqlTransaction;
+    unsafe {
+        tx_safe = cast_pointer!(tx, PgsqlTransaction);
+    }
+    return tx_safe.events;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_pgsql_state_get_event_info(
+    _event_name: *const std::os::raw::c_char, _event_id: *mut std::os::raw::c_int,
+    _event_type: *mut core::AppLayerEventType,
+) -> std::os::raw::c_int {
+    return -1;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_pgsql_state_get_event_info_by_id(
+    _event_id: std::os::raw::c_int, _event_name: *mut *const std::os::raw::c_char,
+    _event_type: *mut core::AppLayerEventType,
+) -> i8 {
+    return -1;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_pgsql_state_get_tx_iterator(
+    _ipproto: u8, _alproto: AppProto, state: *mut std::os::raw::c_void, min_tx_id: u64,
+    _max_tx_id: u64, istate: &mut u64,
+) -> applayer::AppLayerGetTxIterTuple {
+    let state_safe: &mut PgsqlState;
+    unsafe {
+        state_safe = cast_pointer!(state, PgsqlState);
+    }
+    match state_safe.tx_iterator(min_tx_id, istate) {
+        Some((tx, out_tx_id, has_next)) => {
+            let c_tx = tx as *const _ as *mut _;
+            let ires = applayer::AppLayerGetTxIterTuple::with_values(c_tx, out_tx_id, has_next);
+            return ires;
+        }
+        None => {
+            return applayer::AppLayerGetTxIterTuple::not_found();
+        }
+    }
+}
+
+export_tx_data_get!(rs_pgsql_get_tx_data, PgsqlTransaction);
+
+// Parser name as a C style string.
+const PARSER_NAME: &'static [u8] = b"pgsql\0";
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_pgsql_register_parser() {
+    let default_port = CString::new("[5432]").unwrap();
+    let parser = RustParser {
+        name: PARSER_NAME.as_ptr() as *const std::os::raw::c_char,
+        default_port: default_port.as_ptr(),
+        ipproto: IPPROTO_TCP,
+        probe_ts: Some(rs_pgsql_probing_parser_ts),
+        probe_tc: Some(rs_pgsql_probing_parser_tc),
+        min_depth: 0,
+        max_depth: 16,
+        state_new: rs_pgsql_state_new,
+        state_free: rs_pgsql_state_free,
+        tx_free: rs_pgsql_state_tx_free,
+        parse_ts: rs_pgsql_parse_request,
+        parse_tc: rs_pgsql_parse_response,
+        get_tx_count: rs_pgsql_state_get_tx_count,
+        get_tx: rs_pgsql_state_get_tx,
+        tx_comp_st_ts: PgsqlTransactionState::ResponseDone as i32,
+        tx_comp_st_tc: PgsqlTransactionState::ResponseDone as i32,
```

Comment:
Unrelated to your code: Wondering if this should be a u8 in the table as we seem to be checking it against 0 and other status codes < 255 :thinking: